### DISCUSSION TITLE
Difficulty linking Bluetooth 3.0 Keyboard

### DISCUSSION THREAD

**Discussion by Kayak555** [darthcloud/BlueRetro#518] [General]
I can tell that the Blue Retro (GC) sees the keyboard, because if I try to link it first, and then a controller, the controller will go to port 2, however the keyboard never fully syncs and the port does not light up. If you have anything I can try let me know. Otherwise I'll have to buy a new KB.

**Comment by darthcloud** [comment]
Can you try to get me a Bluetooth trace of that keyboard on win10?

Instruction here:
https://github.com/darthcloud/BlueRetro/wiki/Bluetooth-HCI-trace-with-Win10

**Reply by Kayak555** [reply]
[Bluetooth 3.0 Keyboard.zip](https://github.com/darthcloud/BlueRetro/files/10385086/Bluetooth.3.0.Keyboard.zip)

Here you go, if you need any additional information let me know. I started on the top left, top row, went right until every key was hit, then started on the left side on the next row down.

**Reply by darthcloud** [reply]
Are you aware that this Keyboard use PIN code? Is it something you had to set yourself? Or the keyboard always use the same?

What is the PIN?

**Reply by Kayak555** [reply]
Pin differs each time I believe. When I bought it, I didn't know that. Windows asks for a random pairing key. I can confirm when I'm home from work tonight.

**Reply by Kayak555** [reply]
Ok, I confirm, windows asks for an 8 digit key then to hit enter at the end. Android asks for a 6 digit key then to hit enter at the end. Seems like it'll probably be incompatible with the BlueRetro. It's not a big deal.

**Reply by darthcloud** [reply]
No its something I would to support, it ask to type a specific one? or it let youtype anything as long the length is ok?

**Reply by Kayak555** [reply]
It asks for a specific code 8 digit code each time you go to pair it for the first time.

**Reply by darthcloud** [reply]
when the led stop pulsing in the port while making a connection attempt, what happen of you type 0000 on the keyboard at that moment does the port LED become on?

**Reply by Kayak555** [reply]
So, I was actually able to hit 0000 and it links and the LED lights up, but the keyboard won't type anything.

**Reply by darthcloud** [reply]
Did you configure BluerRetro to be a KB in the config?
See https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-BLE-Web-Config-User-Manual#82---keyboard-config-example

I never tried B in port one. I use controller as P1 nd KB as P2 for PSO+ and Action Replay.

**Reply by Kayak555** [reply]
Yes, I've done port 2 as a keyboard already. I set up a different keyboard over the weekend in port 2 and that works just fine.

**Reply by darthcloud** [reply]
ok just to make sure its a kb problem and not a port things can you try that one setup for port 2.?

**Reply by Kayak555** [reply]
I just confirmed that the other keyboard works in port 2, while the other one does not.

**Comment by darthcloud** [comment]
I simulated the keyboard on BlueRetro base on the trace:
```
# 000000 05 01 09 06 A1 01 85 01 75 01 95 08 05 07 19 E0
# 000010 29 E7 15 00 25 01 81 02 95 01 75 08 81 03 95 05
# 000020 75 01 05 08 19 01 29 05 91 02 95 01 75 03 91 03
# 000030 95 06 75 08 15 00 26 FF 00 05 07 19 00 29 FF 81
# 000040 00 C0 05 0C 09 01 A1 01 85 02 15 00 25 01 75 01
# 000050 95 16 0A B1 01 0A 23 02 0A AE 01 0A 8A 01 09 40
# 000060 09 6F 0A 21 02 09 B6 09 CD 09 B5 09 E2 09 EA 09
# 000070 E9 09 30 0A 83 01 0A 24 02 0A 06 03 0A 08 03 0A
# 000080 01 03 0A 83 01 0A 0A 03 09 70 81 02 95 01 75 02
# 000090 81 03 C0
# 1 07E0 0 8 0700 16 8, 0700 24 8, 0700 32 8, 0700 40 8, 0700 48 8, 0700 56 8, rtype: 0 dtype: 0 sub: 0
# 2 0C1B1 0 22 rtype: 3 dtype: -1 sub: 0
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
R0 07E0: 00, R0 0700: 29, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00, R0 0700: 00,
```

HID descriptor parse fine
HID Report are parsed fine as well

So I guess it might not even reach that point, I guess the L2CAP connection fail somehow but without the Keyboard on hand it's not something I can debug remotely unfortunately.

**Comment by Kayak555** [comment]
It's not a huge loss, I can figure out another use for it eventually. Thanks for helping though, I appreciate it.
